datafusion/expr/src/logical_plan/plan.rs

```diff
@@ -603,6 +604,73 @@ impl LogicalPlan {
         Ok(new_plan)
     }
 
+    /// Walk the logical plan, find any `PlaceHolder` tokens, and return a map of their IDs and DataTypes
+    pub fn get_parameter_types(
+        &self,
+    ) -> Result<HashMap<String, Option<DataType>>, DataFusionError> {
+        struct ParamTypeVisitor {
+            param_types: HashMap<String, Option<DataType>>,
+        }
+
+        struct ExprParamTypeVisitor {
+            param_types: HashMap<String, Option<DataType>>,
+        }
+
+        impl ExpressionVisitor for ExprParamTypeVisitor {
+            fn pre_visit(
+                mut self,
+                expr: &Expr,
+            ) -> datafusion_common::Result<Recursion<Self>>
+            where
+                Self: ExpressionVisitor,
+            {
+                if let Expr::Placeholder { id, data_type } = expr {
+                    let _ = self.param_types.insert(id.clone(), data_type.clone());
+                }
+                Ok(Recursion::Continue(self))
+            }
+        }
+
+        impl PlanVisitor for ParamTypeVisitor {
```

Comment:
I suspect you could  `impl PlanVisitor` and `impl ExprVisitor` for `ParamTypeVisitor` rather than having multiple visitors and having to copy items from the hash maps

Reply:
I can, but it doesn't seem to simplify things very much, since the signature of `ExprVisitable` is awkward requiring ownership to be transferred around vs mutation:

```
fn accept<V: ExpressionVisitor>(&self, visitor: V) -> Result<V> {
```

Since I only have a mutable reference to `self` in `pre_visit()` I can't re-use a single instance I don't think:

```
fn pre_visit(&mut self, plan: &LogicalPlan) -> Result<bool, Self::Error> {
```